<center><h1> ETI195 - Etica para Ciencia de datos y Estadistica </h1><center>
<center><h2> Lab Sumativo 2: Fairness <h2></center>

## Contexto:

El set de datos llamado **"adult dataset"** o **"census income dataset"** contiene datos demográficos y financieros de personas, como su edad, nivel educativo, estado civil, ocupación, relación familiar, entre otros. Su principal finalidad es determinar si una persona percibe un ingreso anual de más o menos de $50,000, siendo así un problema de clasificación binaria. Este conjunto de datos se considera delicado debido a que contiene información personal e ingresos, por lo que regularmente se emplea para enseñar y poner en práctica procesos de análisis de datos de forma ética y responsable.

En la presente tarea se entrega la versión de este dataset disponible en [AIF360](https://aif360.readthedocs.io/en/stable/modules/generated/aif360.datasets.AdultDataset.html#aif360.datasets.AdultDataset), con algunas modificaciones. El objetivo es evaluar la implementación de los tooltkits Aequitas y AIF360 en el contexto de este dataset.
## Instrucciones:

- La siguiente tarea es **individual**.
- Fecha de entrega: Lunes 13 de mayo, en Canvas.

In [1]:
import pandas as pd
import numpy as np


In [2]:

from sklearn.linear_model import LogisticRegressionCV, LogisticRegression
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.metrics import accuracy_score, confusion_matrix, recall_score, precision_score, f1_score, roc_auc_score, roc_curve

In [50]:
from aif360.sklearn.datasets import fetch_german
from aif360.algorithms.preprocessing import Reweighing
from aif360.metrics import BinaryLabelDatasetMetric
from aif360.datasets import BinaryLabelDataset
from aif360.algorithms.postprocessing import EqOddsPostprocessing
from aif360.metrics import ClassificationMetric


In [4]:
df = pd.read_csv('adult_data.csv')
df.head()

age  workclass     education  education-num      marital-status  \
0  25.0    Private          11th            7.0       Never-married   
1  38.0    Private       HS-grad            9.0  Married-civ-spouse   
2  28.0  Local-gov    Assoc-acdm           12.0  Married-civ-spouse   
3  44.0    Private  Some-college           10.0  Married-civ-spouse   
4  34.0    Private          10th            6.0       Never-married   

          occupation   relationship   race   sex  capital-gain  capital-loss  \
0  Machine-op-inspct      Own-child  Black  Male           0.0           0.0   
1    Farming-fishing        Husband  White  Male           0.0           0.0   
2    Protective-serv        Husband  White  Male           0.0           0.0   
3  Machine-op-inspct        Husband  Black  Male        7688.0           0.0   
4      Other-service  Not-in-family  White  Male           0.0           0.0   

   hours-per-week native-country income  
0            40.0  United-States  <=50K  
1            50.0  United-States  <=50K  
2            40.0  United-States   >50K  
3            40.0  United-States   >50K  
4            30.0  United-States  <=50K

### Pregunta 1: Realice el pre-procesamiento de los datos e implemente un modelo de clasificación (Regresión Logística). Obtenga el accuracy del modelo sobre los datos de testeo (1 pto).

#### Indicaciones adicionales:

- Recuerde utilizar random_state o seeds según sea necesario para la reproducibilidad de sus resultados.

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45222 entries, 0 to 45221
Data columns (total 14 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   age             45222 non-null  float64
 1   workclass       45222 non-null  object 
 2   education       45222 non-null  object 
 3   education-num   45222 non-null  float64
 4   marital-status  45222 non-null  object 
 5   occupation      45222 non-null  object 
 6   relationship    45222 non-null  object 
 7   race            45222 non-null  object 
 8   sex             45222 non-null  object 
 9   capital-gain    45222 non-null  float64
 10  capital-loss    45222 non-null  float64
 11  hours-per-week  45222 non-null  float64
 12  native-country  45222 non-null  object 
 13  income          45222 non-null  object 
dtypes: float64(5), object(9)
memory usage: 4.8+ MB


In [6]:
# Binarizacion de algunas variables categoricas para poder trabajar con ellas en el modelo de regresion logistica 
df['sex'] = df['sex'].map({'Male': 1, 'Female': 0})
df['income'] = df['income'].map({'<=50K': 0, '>50K': 1})

df

age     workclass     education  education-num      marital-status  \
0      25.0       Private          11th            7.0       Never-married   
1      38.0       Private       HS-grad            9.0  Married-civ-spouse   
2      28.0     Local-gov    Assoc-acdm           12.0  Married-civ-spouse   
3      44.0       Private  Some-college           10.0  Married-civ-spouse   
4      34.0       Private          10th            6.0       Never-married   
...     ...           ...           ...            ...                 ...   
45217  27.0       Private    Assoc-acdm           12.0  Married-civ-spouse   
45218  40.0       Private       HS-grad            9.0  Married-civ-spouse   
45219  58.0       Private       HS-grad            9.0             Widowed   
45220  22.0       Private       HS-grad            9.0       Never-married   
45221  52.0  Self-emp-inc       HS-grad            9.0  Married-civ-spouse   

              occupation   relationship   race  sex  capital-gain  \
0      Machine-op-inspct      Own-child  Black    1           0.0   
1        Farming-fishing        Husband  White    1           0.0   
2        Protective-serv        Husband  White    1           0.0   
3      Machine-op-inspct        Husband  Black    1        7688.0   
4          Other-service  Not-in-family  White    1           0.0   
...                  ...            ...    ...  ...           ...   
45217       Tech-support           Wife  White    0           0.0   
45218  Machine-op-inspct        Husband  White    1           0.0   
45219       Adm-clerical      Unmarried  White    0           0.0   
45220       Adm-clerical      Own-child  White    1           0.0   
45221    Exec-managerial           Wife  White    0       15024.0   

       capital-loss  hours-per-week native-country  income  
0               0.0            40.0  United-States       0  
1               0.0            50.0  United-States       0  
2               0.0            40.0  United-States       1  
3               0.0            40.0  United-States       1  
4               0.0            30.0  United-States       0  
...             ...             ...            ...     ...  
45217           0.0            38.0  United-States       0  
45218           0.0            40.0  United-States       1  
45219           0.0            40.0  United-States       0  
45220           0.0            20.0  United-States       0  
45221           0.0            40.0  United-States       1  

[45222 rows x 14 columns]

In [7]:
# Separamos x e y
X = df.loc[:, df.columns != 'income']
y = df.loc[:, df.columns == 'income']


In [8]:
df1 = X.copy()


In [9]:
# Obtenemos variables dummies
ignore = ['sex']

for col in df1.select_dtypes(include='object').columns:
    if col in ignore:
        continue
    else:
        dummies = pd.get_dummies(X[col])
        X = pd.concat([X, dummies], axis=1).drop(columns=[col])
X.head()

age  education-num  sex  capital-gain  capital-loss  hours-per-week  \
0  25.0            7.0    1           0.0           0.0            40.0   
1  38.0            9.0    1           0.0           0.0            50.0   
2  28.0           12.0    1           0.0           0.0            40.0   
3  44.0           10.0    1        7688.0           0.0            40.0   
4  34.0            6.0    1           0.0           0.0            30.0   

   Federal-gov  Local-gov  Private  Self-emp-inc  ...  Portugal  Puerto-Rico  \
0        False      False     True         False  ...     False        False   
1        False      False     True         False  ...     False        False   
2        False       True    False         False  ...     False        False   
3        False      False     True         False  ...     False        False   
4        False      False     True         False  ...     False        False   

   Scotland  South  Taiwan  Thailand  Trinadad&Tobago  United-States  Vietnam  \
0     False  False   False     False            False           True    False   
1     False  False   False     False            False           True    False   
2     False  False   False     False            False           True    False   
3     False  False   False     False            False           True    False   
4     False  False   False     False            False           True    False   

   Yugoslavia  
0       False  
1       False  
2       False  
3       False  
4       False  

[5 rows x 102 columns]

In [10]:
# Cambiamos false por 0 y true por 1
X = X.replace({False: 0, True: 1})
X.head()

age  education-num  sex  capital-gain  capital-loss  hours-per-week  \
0  25.0            7.0    1           0.0           0.0            40.0   
1  38.0            9.0    1           0.0           0.0            50.0   
2  28.0           12.0    1           0.0           0.0            40.0   
3  44.0           10.0    1        7688.0           0.0            40.0   
4  34.0            6.0    1           0.0           0.0            30.0   

   Federal-gov  Local-gov  Private  Self-emp-inc  ...  Portugal  Puerto-Rico  \
0            0          0        1             0  ...         0            0   
1            0          0        1             0  ...         0            0   
2            0          1        0             0  ...         0            0   
3            0          0        1             0  ...         0            0   
4            0          0        1             0  ...         0            0   

   Scotland  South  Taiwan  Thailand  Trinadad&Tobago  United-States  Vietnam  \
0         0      0       0         0                0              1        0   
1         0      0       0         0                0              1        0   
2         0      0       0         0                0              1        0   
3         0      0       0         0                0              1        0   
4         0      0       0         0                0              1        0   

   Yugoslavia  
0           0  
1           0  
2           0  
3           0  
4           0  

[5 rows x 102 columns]

In [11]:
# Datos de entrenamiento y test 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=908)


In [12]:
# Sacamos la edad
X_train.drop(columns=['age'], inplace=True)
X_test.drop(columns=['age'], inplace=True)


In [13]:
df_train = X_train.copy()
df_train['income'] = y_train


In [14]:
# Discriminacion por sexo 

privileged_groups = [{'sex': 1}]
unprivileged_groups = [{'sex': 0}]


In [15]:
# instanciamos binarylabeldataset
attribute_params = dict(protected_attribute_names=['sex'], label_names=['income'])
dt_train = BinaryLabelDataset(df=df_train, **attribute_params)


In [16]:
# Finalmente, la instancia de BinaryLabelDatasetMetric con la edad como atributo protegido.

metric_age_train = BinaryLabelDatasetMetric(dt_train,
                                             unprivileged_groups=unprivileged_groups,
                                             privileged_groups=privileged_groups)

In [17]:
print(f'Train Statistical Parity Difference (age): {metric_age_train.mean_difference()}')

Train Statistical Parity Difference (age): -0.1975830166220287


In [18]:
lr = LogisticRegressionCV(solver='liblinear', cv=10, random_state=908, max_iter=1000)
lr.fit(X_train, np.ravel(y_train))

LogisticRegressionCV(cv=10, max_iter=1000, random_state=908, solver='liblinear')

In [19]:
y_pred_proba = lr.predict_proba(X_test)[:, 1]
y_pred = y_pred_proba >= 0.5

print(f'Accuracy: {accuracy_score(y_test, y_pred)}')

print(f"Recall: {recall_score(y_test, y_pred)}")
print(f"F1: {f1_score(y_test, y_pred)}")


Accuracy: 0.8437810945273632
Recall: 0.5729632945389436
F1: 0.6443493581676315


### Pregunta 2: Evalúe las clasificaciones del modelo de la pregunta anterior a través de las métricas de Fairness facilitadas en Aequitas.

In [20]:
from aequitas.group import Group

In [21]:
# Creare una columna de age_cat,  con greater than 45, 25- 45 y Less than 45
df['age_cat'] = pd.cut(df['age'], bins=[0, 25, 45, 100], labels=['Less than 25', '25-45', 'Greater than 45'])

df

age     workclass     education  education-num      marital-status  \
0      25.0       Private          11th            7.0       Never-married   
1      38.0       Private       HS-grad            9.0  Married-civ-spouse   
2      28.0     Local-gov    Assoc-acdm           12.0  Married-civ-spouse   
3      44.0       Private  Some-college           10.0  Married-civ-spouse   
4      34.0       Private          10th            6.0       Never-married   
...     ...           ...           ...            ...                 ...   
45217  27.0       Private    Assoc-acdm           12.0  Married-civ-spouse   
45218  40.0       Private       HS-grad            9.0  Married-civ-spouse   
45219  58.0       Private       HS-grad            9.0             Widowed   
45220  22.0       Private       HS-grad            9.0       Never-married   
45221  52.0  Self-emp-inc       HS-grad            9.0  Married-civ-spouse   

              occupation   relationship   race  sex  capital-gain  \
0      Machine-op-inspct      Own-child  Black    1           0.0   
1        Farming-fishing        Husband  White    1           0.0   
2        Protective-serv        Husband  White    1           0.0   
3      Machine-op-inspct        Husband  Black    1        7688.0   
4          Other-service  Not-in-family  White    1           0.0   
...                  ...            ...    ...  ...           ...   
45217       Tech-support           Wife  White    0           0.0   
45218  Machine-op-inspct        Husband  White    1           0.0   
45219       Adm-clerical      Unmarried  White    0           0.0   
45220       Adm-clerical      Own-child  White    1           0.0   
45221    Exec-managerial           Wife  White    0       15024.0   

       capital-loss  hours-per-week native-country  income          age_cat  
0               0.0            40.0  United-States       0     Less than 25  
1               0.0            50.0  United-States       0            25-45  
2               0.0            40.0  United-States       1            25-45  
3               0.0            40.0  United-States       1            25-45  
4               0.0            30.0  United-States       0            25-45  
...             ...             ...            ...     ...              ...  
45217           0.0            38.0  United-States       0            25-45  
45218           0.0            40.0  United-States       1            25-45  
45219           0.0            40.0  United-States       0  Greater than 45  
45220           0.0            20.0  United-States       0     Less than 25  
45221           0.0            40.0  United-States       1  Greater than 45  

[45222 rows x 15 columns]

In [22]:
# Seleccionamos las columnas necesarias

df_aequitas = df[['income', 'sex', 'race', 'age_cat', 'education-num']]
df_aequitas

income  sex   race          age_cat  education-num
0           0    1  Black     Less than 25            7.0
1           0    1  White            25-45            9.0
2           1    1  White            25-45           12.0
3           1    1  Black            25-45           10.0
4           0    1  White            25-45            6.0
...       ...  ...    ...              ...            ...
45217       0    0  White            25-45           12.0
45218       1    1  White            25-45            9.0
45219       0    0  White  Greater than 45            9.0
45220       0    1  White     Less than 25            9.0
45221       1    0  White  Greater than 45            9.0

[45222 rows x 5 columns]

In [23]:
df_aequitas.rename(columns={'sex': 'score', 'income': 'label_value'}, inplace=True)


C:\Users\Poble\AppData\Local\Temp\ipykernel_4940\21422225.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_aequitas.rename(columns={'sex': 'score', 'income': 'label_value'}, inplace=True)


In [24]:
df_aequitas.head()

label_value  score   race       age_cat  education-num
0            0      1  Black  Less than 25            7.0
1            0      1  White         25-45            9.0
2            1      1  White         25-45           12.0
3            1      1  Black         25-45           10.0
4            0      1  White         25-45            6.0

In [25]:

df_aequitas.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45222 entries, 0 to 45221
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype   
---  ------         --------------  -----   
 0   label_value    45222 non-null  int64   
 1   score          45222 non-null  int64   
 2   race           45222 non-null  object  
 3   age_cat        45222 non-null  category
 4   education-num  45222 non-null  float64 
dtypes: category(1), float64(1), int64(2), object(1)
memory usage: 1.4+ MB


In [26]:
# Cambiamos el tipo de la columna age_cat a object 
df_aequitas['age_cat'] = df_aequitas['age_cat'].astype('object')
df_aequitas['education-num'] = df_aequitas['education-num'].astype('object')

df_aequitas.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45222 entries, 0 to 45221
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   label_value    45222 non-null  int64 
 1   score          45222 non-null  int64 
 2   race           45222 non-null  object
 3   age_cat        45222 non-null  object
 4   education-num  45222 non-null  object
dtypes: int64(2), object(3)
memory usage: 1.7+ MB


C:\Users\Poble\AppData\Local\Temp\ipykernel_4940\3947863477.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_aequitas['age_cat'] = df_aequitas['age_cat'].astype('object')
C:\Users\Poble\AppData\Local\Temp\ipykernel_4940\3947863477.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_aequitas['education-num'] = df_aequitas['education-num'].astype('object')


In [27]:
group = Group()
xtab, _ = group.get_crosstabs(df_aequitas)
xtab.head()

model_id score_threshold      k attribute_name     attribute_value  \
0         0      binary 0/1  30527           race  Amer-Indian-Eskimo   
1         0      binary 0/1  30527           race  Asian-Pac-Islander   
2         0      binary 0/1  30527           race               Black   
3         0      binary 0/1  30527           race               Other   
4         0      binary 0/1  30527           race               White   

   accuracy       tpr       tnr       for       fdr  ...     pprev     fp  \
0  0.439080  0.735849  0.397906  0.084337  0.855019  ...  0.618391    230   
1  0.518035  0.823848  0.397216  0.149083  0.649366  ...  0.665388    563   
2  0.559603  0.764045  0.530049  0.060461  0.809701  ...  0.507096   1736   
3  0.433428  0.800000  0.379870  0.071429  0.841410  ...  0.643059    191   
4  0.493021  0.857451  0.363396  0.122444  0.676092  ...  0.694548  18268   

     fn     tn    tp  group_label_pos  group_label_neg  group_size  \
0    14    152    39               53              382         435   
1    65    371   304              369              934        1303   
2   126   1958   408              534             3694        4228   
3     9    117    36               45              308         353   
4  1455  10428  8752            10207            28696       38903   

   total_entities      prev  
0           45222  0.121839  
1           45222  0.283193  
2           45222  0.126301  
3           45222  0.127479  
4           45222  0.262371  

[5 rows x 27 columns]

In [28]:
xtab.columns

Index(['model_id', 'score_threshold', 'k', 'attribute_name', 'attribute_value',
       'accuracy', 'tpr', 'tnr', 'for', 'fdr', 'fpr', 'fnr', 'npv',
       'precision', 'pp', 'pn', 'ppr', 'pprev', 'fp', 'fn', 'tn', 'tp',
       'group_label_pos', 'group_label_neg', 'group_size', 'total_entities',
       'prev'],
      dtype='object')

In [29]:
absolute_metrics = group.list_absolute_metrics(xtab)
absolute_metrics

['accuracy',
 'tpr',
 'tnr',
 'for',
 'fdr',
 'fpr',
 'fnr',
 'npv',
 'precision',
 'ppr',
 'pprev',
 'prev']

In [30]:
absolute_xtab = xtab[['attribute_name', 'attribute_value'] + absolute_metrics]
absolute_xtab

attribute_name     attribute_value  accuracy       tpr       tnr       for  \
0            race  Amer-Indian-Eskimo  0.439080  0.735849  0.397906  0.084337   
1            race  Asian-Pac-Islander  0.518035  0.823848  0.397216  0.149083   
2            race               Black  0.559603  0.764045  0.530049  0.060461   
3            race               Other  0.433428  0.800000  0.379870  0.071429   
4            race               White  0.493021  0.857451  0.363396  0.122444   
5         age_cat               25-45  0.481271  0.834465  0.351380  0.147668   
6         age_cat     Greater than 45  0.563959  0.878023  0.390212  0.147436   
7         age_cat        Less than 25  0.446511  0.717791  0.441169  0.012439   
8   education-num                 1.0  0.291667  1.000000  0.281690  0.000000   
9   education-num                 2.0  0.288288  1.000000  0.261682  0.000000   
10  education-num                 3.0  0.273942  0.818182  0.245902  0.036697   
11  education-num                 4.0  0.303767  0.981818  0.255208  0.005076   
12  education-num                 5.0  0.318047  0.868421  0.285266  0.026738   
13  education-num                 6.0  0.381848  0.963415  0.340053  0.007673   
14  education-num                 7.0  0.384805  0.876404  0.356209  0.019784   
15  education-num                 8.0  0.377816  0.953488  0.331461  0.011173   
16  education-num                 9.0  0.436785  0.871689  0.351823  0.066509   
17  education-num                10.0  0.520457  0.861307  0.434695  0.074313   
18  education-num                11.0  0.513017  0.827381  0.404124  0.128889   
19  education-num                12.0  0.546118  0.798995  0.455365  0.136752   
20  education-num                13.0  0.597886  0.845500  0.418716  0.210730   
21  education-num                14.0  0.655529  0.803302  0.471900  0.341220   
22  education-num                15.0  0.754140  0.895270  0.321244  0.500000   
23  education-num                16.0  0.700368  0.842105  0.310345  0.583333   

         fdr       fpr       fnr       npv  precision       ppr     pprev  \
0   0.855019  0.602094  0.264151  0.915663   0.144981  0.008812  0.618391   
1   0.649366  0.602784  0.176152  0.850917   0.350634  0.028401  0.665388   
2   0.809701  0.469951  0.235955  0.939539   0.190299  0.070233  0.507096   
3   0.841410  0.620130  0.200000  0.928571   0.158590  0.007436  0.643059   
4   0.676092  0.636604  0.142549  0.877556   0.323908  0.885118  0.694548   
5   0.678825  0.648620  0.165535  0.852332   0.321175  0.538834  0.698590   
6   0.556615  0.609788  0.121977  0.852564   0.443385  0.305795  0.705327   
7   0.975332  0.558831  0.282209  0.987561   0.024668  0.155371  0.561900   
8   0.980769  0.718310  0.000000  1.000000   0.019231  0.001703  0.722222   
9   0.951807  0.738318  0.000000  1.000000   0.048193  0.005438  0.747748   
10  0.947059  0.754098  0.181818  0.963303   0.052941  0.011138  0.757238   
11  0.913738  0.744792  0.018182  0.994924   0.086262  0.020506  0.760632   
12  0.932515  0.714734  0.131579  0.973262   0.067485  0.016019  0.723373   
13  0.905048  0.659947  0.036585  0.992327   0.094952  0.027255  0.680294   
14  0.926623  0.643791  0.123596  0.980216   0.073377  0.034822  0.656578   
15  0.896985  0.668539  0.046512  0.988827   0.103015  0.013038  0.689775   
16  0.791938  0.648177  0.128311  0.933491   0.208062  0.331575  0.684705   
17  0.722878  0.565305  0.138693  0.925687   0.277122  0.202608  0.624811   
18  0.675234  0.595876  0.172619  0.871111   0.324766  0.042061  0.655436   
19  0.655098  0.544635  0.201005  0.863248   0.344902  0.030203  0.611812   
20  0.487214  0.581284  0.154500  0.789270   0.512786  0.171651  0.692206   
21  0.345996  0.528100  0.196698  0.658780   0.654004  0.056049  0.680589   
22  0.198185  0.678756  0.104730  0.500000   0.801815  0.021653  0.842038   
23  0.229358  0.689655  0.157895  0.416667   0.770642  0.014282  0.801471   

        prev  
0   0.121839  
1   0.283193  
2   0.

In [31]:
xtab.shape

(24, 27)

### Pregunta 3: Aplique el algoritmo EqOddsPostProcessing sobre las predicciones del modelo de la pregunta 1. Obtenga las predicciones modificadas y evalúelas. ¿Qué cambios hubo en las métricas de Fairness? (2 ptos).

#### Indicaciones adicionales:

- Obtenga tanto las intancias de BinaryLabelDatasetMetric como de ClassificationMetric y calcule al menos una métrica de Fairness.


In [32]:
X_test

education-num  sex  capital-gain  capital-loss  hours-per-week  \
24938           13.0    1           0.0           0.0            30.0   
22771           13.0    1        4386.0           0.0            47.0   
6957            14.0    0           0.0           0.0            45.0   
23546            9.0    1           0.0           0.0            40.0   
18142            9.0    1           0.0           0.0            40.0   
...              ...  ...           ...           ...             ...   
29934           10.0    0           0.0           0.0            40.0   
10608            9.0    1           0.0           0.0            40.0   
32133            5.0    1           0.0           0.0            35.0   
41809            9.0    1           0.0           0.0            45.0   
18705           10.0    0           0.0           0.0            38.0   

       Federal-gov  Local-gov  Private  Self-emp-inc  Self-emp-not-inc  ...  \
24938            0          0        1             0                 0  ...   
22771            0          0        1             0                 0  ...   
6957             0          1        0             0                 0  ...   
23546            0          0        1             0                 0  ...   
18142            0          0        1             0                 0  ...   
...            ...        ...      ...           ...               ...  ...   
29934            0          0        1             0                 0  ...   
10608            0          0        1             0                 0  ...   
32133            0          0        1             0                 0  ...   
41809            0          0        1             0                 0  ...   
18705            0          0        1             0                 0  ...   

       Portugal  Puerto-Rico  Scotland  South  Taiwan  Thailand  \
24938         0            0         0      0       0         0   
22771         0            0         0      0       0         0   
6957          0            0         0      0       0         0   
23546         0            0         0      0       0         0   
18142         0            0         0      0       0         0   
...         ...          ...       ...    ...     ...       ...   
29934         0            0         0      0       0         0   
10608         0            0         0      0       0         0   
32133         0            0         0      0       0         0   
41809         0            0         0      0       0         0   
18705         0            0         0      0       0         0   

       Trinadad&Tobago  United-States  Vietnam  Yugoslavia  
24938                0              1        0           0  
22771                0              1        0           0  
6957                 0              1        0           0  
23546                0              1        0           0  
18142                0              1        0           0  
...                ...            ...      ...         ...  
29934                0              1        0           0  
10608                0              1        0           0  
32133                0              0        0           0  
41809                0              1        0           0  
18705                0              1        0           0  

[9045 rows x 101 columns]

In [33]:
X_test

education-num  sex  capital-gain  capital-loss  hours-per-week  \
24938           13.0    1           0.0           0.0            30.0   
22771           13.0    1        4386.0           0.0            47.0   
6957            14.0    0           0.0           0.0            45.0   
23546            9.0    1           0.0           0.0            40.0   
18142            9.0    1           0.0           0.0            40.0   
...              ...  ...           ...           ...             ...   
29934           10.0    0           0.0           0.0            40.0   
10608            9.0    1           0.0           0.0            40.0   
32133            5.0    1           0.0           0.0            35.0   
41809            9.0    1           0.0           0.0            45.0   
18705           10.0    0           0.0           0.0            38.0   

       Federal-gov  Local-gov  Private  Self-emp-inc  Self-emp-not-inc  ...  \
24938            0          0        1             0                 0  ...   
22771            0          0        1             0                 0  ...   
6957             0          1        0             0                 0  ...   
23546            0          0        1             0                 0  ...   
18142            0          0        1             0                 0  ...   
...            ...        ...      ...           ...               ...  ...   
29934            0          0        1             0                 0  ...   
10608            0          0        1             0                 0  ...   
32133            0          0        1             0                 0  ...   
41809            0          0        1             0                 0  ...   
18705            0          0        1             0                 0  ...   

       Portugal  Puerto-Rico  Scotland  South  Taiwan  Thailand  \
24938         0            0         0      0       0         0   
22771         0            0         0      0       0         0   
6957          0            0         0      0       0         0   
23546         0            0         0      0       0         0   
18142         0            0         0      0       0         0   
...         ...          ...       ...    ...     ...       ...   
29934         0            0         0      0       0         0   
10608         0            0         0      0       0         0   
32133         0            0         0      0       0         0   
41809         0            0         0      0       0         0   
18705         0            0         0      0       0         0   

       Trinadad&Tobago  United-States  Vietnam  Yugoslavia  
24938                0              1        0           0  
22771                0              1        0           0  
6957                 0              1        0           0  
23546                0              1        0           0  
18142                0              1        0           0  
...                ...            ...      ...         ...  
29934                0              1        0           0  
10608                0              1        0           0  
32133                0              0        0           0  
41809                0              1        0           0  
18705                0              1        0           0  

[9045 rows x 101 columns]

In [34]:
df_test = X_test.copy()
df_test['income'] = y_pred

In [35]:
def get_aif_metrics(attr, df):

  privileged_groups = [{attr: 1}]
  unprivileged_groups = [{attr: 0}]

  attributes_params = dict(protected_attribute_names=[attr], label_names=['income'])
  dt = BinaryLabelDataset(df=df, **attributes_params)

  metric = BinaryLabelDatasetMetric(dt,
                                    unprivileged_groups=unprivileged_groups,
                                    privileged_groups=privileged_groups)

  return metric

In [36]:
metric_sex_test = get_aif_metrics('sex', df_test)
print(f'Test Statistical Parity Difference (sex): {metric_sex_test.mean_difference()}')
print(f'Test Disparate Impact (sex): {metric_sex_test.disparate_impact()}')

Test Statistical Parity Difference (sex): -0.17033771845504564
Test Disparate Impact (sex): 0.3080170084050281


In [37]:
metric_race_test = get_aif_metrics('White', df_test)
print(f'Test Statistical Parity Difference (race) : {metric_race_test.mean_difference()}')  
print(f'Test Disparate Imapct (race): {metric_race_test.disparate_impact()}')


Test Statistical Parity Difference (race) : -0.07310620725254871
Test Disparate Imapct (race): 0.6394016549968173


In [54]:
def get_aif_metrics_2(attr, df, label_names):

  privileged_groups = [{attr: 1}]
  unprivileged_groups = [{attr: 0}]

  attributes_params = dict(protected_attribute_names=[attr], label_names=['income'])
  dt = BinaryLabelDataset(df=df, **attributes_params)

  metric = BinaryLabelDatasetMetric(dt,
                                    unprivileged_groups=unprivileged_groups,
                                    privileged_groups=privileged_groups)

  return metric

education-num  sex  capital-gain  capital-loss  hours-per-week  \
28441            9.0    0           0.0           0.0            40.0   
33791            9.0    1           0.0        1579.0            40.0   
21441           13.0    1           0.0           0.0            40.0   
9154            14.0    0           0.0        1902.0            45.0   
23791           10.0    1           0.0           0.0            40.0   
...              ...  ...           ...           ...             ...   
6456             9.0    1           0.0           0.0            40.0   
11196           11.0    0           0.0        1564.0            72.0   
25904           12.0    0           0.0           0.0            40.0   
35322           13.0    1           0.0           0.0            50.0   
18538           10.0    1           0.0           0.0            40.0   

       Federal-gov  Local-gov  Private  Self-emp-inc  Self-emp-not-inc  ...  \
28441            0          0        1             0                 0  ...   
33791            0          0        1             0                 0  ...   
21441            0          0        1             0                 0  ...   
9154             0          0        1             0                 0  ...   
23791            0          0        1             0                 0  ...   
...            ...        ...      ...           ...               ...  ...   
6456             0          0        1             0                 0  ...   
11196            0          1        0             0                 0  ...   
25904            0          1        0             0                 0  ...   
35322            0          0        1             0                 0  ...   
18538            0          0        1             0                 0  ...   

       Puerto-Rico  Scotland  South  Taiwan  Thailand  Trinadad&Tobago  \
28441            0         0      0       0         0                0   
33791            0         0      0       0         0                0   
21441            0         0      0       0         0                0   
9154             0         0      0       0         0                0   
23791            0         0      0       0         0                0   
...            ...       ...    ...     ...       ...              ...   
6456             0         0      0       0         0                0   
11196            0         0      0       0         0                0   
25904            0         0      0       0         0                0   
35322            0         0      0       0         0                0   
18538            0         0      0       0         0                0   

       United-States  Vietnam  Yugoslavia  income  
28441              1        0           0       0  
33791              1        0           0       0  
21441              1        0           0       1  
9154               1        0           0       1  
23791              1        0           0       1  
...              ...      ...         ...     ...  
6456               1        0           0       0  
11196              1        0           0       1  
25904              1        0           0       0  
35322              1        0           0       1  
18538              1        0           0       0  

[36177 rows x 102 columns]

In [39]:
dt_train.instance_weights

array([1., 1., 1., ..., 1., 1., 1.])

### Pregunta 4: Aplique el algoritmo de Reweighing sobre los datos de entrenamiento. Compare las métricas de Fairness para los datos de entrenamiento antes y después de aplicar Reweighing. Luego, entrene un modelo de Regresión Logística utilizando los datos de entrenamiento modificados y obtenga las métricas de Fairness de las predicciones de dicho modelo. ¿Cómo varió el accuracy del modelo respecto al original y las métricas de fairness? (2 ptos).

#### Indicaciones adicionales:

- Considere 'Black' como grupo no privilegiado para la raza, 'White' como grupo privilegiado para la raza.
- Considere 'Female' como grupo no privilegiado para el sexo, 'Male' como grupo privilegiado para el sexo.


In [40]:
df_train

education-num  sex  capital-gain  capital-loss  hours-per-week  \
28441            9.0    0           0.0           0.0            40.0   
33791            9.0    1           0.0        1579.0            40.0   
21441           13.0    1           0.0           0.0            40.0   
9154            14.0    0           0.0        1902.0            45.0   
23791           10.0    1           0.0           0.0            40.0   
...              ...  ...           ...           ...             ...   
6456             9.0    1           0.0           0.0            40.0   
11196           11.0    0           0.0        1564.0            72.0   
25904           12.0    0           0.0           0.0            40.0   
35322           13.0    1           0.0           0.0            50.0   
18538           10.0    1           0.0           0.0            40.0   

       Federal-gov  Local-gov  Private  Self-emp-inc  Self-emp-not-inc  ...  \
28441            0          0        1             0                 0  ...   
33791            0          0        1             0                 0  ...   
21441            0          0        1             0                 0  ...   
9154             0          0        1             0                 0  ...   
23791            0          0        1             0                 0  ...   
...            ...        ...      ...           ...               ...  ...   
6456             0          0        1             0                 0  ...   
11196            0          1        0             0                 0  ...   
25904            0          1        0             0                 0  ...   
35322            0          0        1             0                 0  ...   
18538            0          0        1             0                 0  ...   

       Puerto-Rico  Scotland  South  Taiwan  Thailand  Trinadad&Tobago  \
28441            0         0      0       0         0                0   
33791            0         0      0       0         0                0   
21441            0         0      0       0         0                0   
9154             0         0      0       0         0                0   
23791            0         0      0       0         0                0   
...            ...       ...    ...     ...       ...              ...   
6456             0         0      0       0         0                0   
11196            0         0      0       0         0                0   
25904            0         0      0       0         0                0   
35322            0         0      0       0         0                0   
18538            0         0      0       0         0                0   

       United-States  Vietnam  Yugoslavia  income  
28441              1        0           0       0  
33791              1        0           0       0  
21441              1        0           0       1  
9154               1        0           0       1  
23791              1        0           0       1  
...              ...      ...         ...     ...  
6456               1        0           0       0  
11196              1        0           0       1  
25904              1        0           0       0  
35322              1        0           0       1  
18538              1        0           0       0  

[36177 rows x 102 columns]

In [41]:
# Aplicaremos el algoritmo de Reweighing para balancear los datos

privileged_groups = [{'White': 1}]
unprivileged_groups = [{'Black': 1}]

privileged_groups_sex= [{'sex': 1}]
unprivileged_groups_sex = [{'sex': 0}]


# Instanciamos BinaryLabelDataset con edad como atributo protegido

attributes_params = dict(protected_attribute_names=['White', 'Black'], label_names=['income'])
dt_train = BinaryLabelDataset(df=df_train, **attributes_params)

# Instanciamos Rw
rw = Reweighing(unprivileged_groups=unprivileged_groups, privileged_groups=privileged_groups)

#Obtenemos los pesos

reweighted = rw.fit_transform(dt_train)



In [42]:
reweighted.instance_weights

array([1.02002944, 1.02002944, 0.9438207 , ..., 0.86305616, 0.9438207 ,
       1.02002944])

In [43]:
reweighted_metrics = BinaryLabelDatasetMetric(reweighted,
                                              unprivileged_groups=unprivileged_groups,
                                              privileged_groups=privileged_groups)


In [44]:
print(f'Reweighted Train Statistical Parity Difference (age): {reweighted_metrics.mean_difference()}')

Reweighted Train Statistical Parity Difference (age): -2.7755575615628914e-17


In [45]:
# Regresion Logistica
lr = LogisticRegressionCV(solver='liblinear', cv=10, random_state=908)

# Agregamos los pesos calculados por el algoritmo de reweighing
lr.fit(X_train, np.ravel(y_train), sample_weight=reweighted.instance_weights)

LogisticRegressionCV(cv=10, random_state=908, solver='liblinear')

Metrica modelo entrenado con datos "reparados"

In [46]:
y_pred_proba = lr.predict_proba(X_test)[:, 1]
y_pred = y_pred_proba >= 0.5

print(f'Accuracy: {accuracy_score(y_test, y_pred)}')
print(f"Recall: {recall_score(y_test, y_pred)}")
print(f"F1: {f1_score(y_test, y_pred)}")

Accuracy: 0.8438916528468767
Recall: 0.5720680393912265
F1: 0.6441532258064516


Sesgo del modelo entrenado con datos "reparados"

In [47]:
df_test_rw = X_test.copy()
df_test_rw['income'] = y_pred


Discriminacion por sexo

In [48]:
metric_sex_test_rw = get_aif_metrics('sex', df_test_rw)
print(f'Reweighted: Test Statistical Parity Difference (sex): {metric_sex_test_rw.mean_difference()}')
print(f'Reweighted: Test Disparate Impact (sex): {metric_sex_test_rw.disparate_impact()}')

Reweighted: Test Statistical Parity Difference (sex): -0.15777283421676597
Reweighted: Test Disparate Impact (sex): 0.34704857164507086


In [49]:
metric_sex_test_rw = get_aif_metrics('White', df_test_rw)
print(f'Reweighted: Test Statistical Parity Difference (race): {metric_sex_test_rw.mean_difference()}')
print(f'Reweighted: Test Disparate Impact (race): {metric_sex_test_rw.disparate_impact()}')

Reweighted: Test Statistical Parity Difference (race): -0.03283219340129909
Reweighted: Test Disparate Impact (race): 0.8328405606657907


Para sex, vario de  de Test Statistical Parity Difference (sex): -0.17033771845504564  y Test Disparate Impact (sex): 0.3080170084050281
Reweighted: Test Statistical Parity Difference (sex): -0.15777283421676597 y Reweighted: Test Disparate Impact (sex): 0.34704857164507086

Para race vario de:
Test Statistical Parity Difference (race) : -0.07310620725254871 y Test Disparate Imapct (race): 0.6394016549968173 a 
Reweighted: Test Statistical Parity Difference (race): -0.03283219340129909 y Reweighted: Test Disparate Impact (race): 0.8328405606657907
